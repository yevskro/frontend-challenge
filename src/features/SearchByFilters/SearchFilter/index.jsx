import React from 'react';
import styled from 'styled-components';

function SearchFilter({
  children,
  filter,
  remove,
  onRemoveFilter,
  onAddFilter,
}) {
  /* a filter that is removeable should only handle an onClick
  on the removing icon.
    a filter that is addable shall not have the remove prop or
    the prop would be set to false. this filter can be clickable
    by the text content
  */
  return (
    <StyledSearchFilter>
      <Text
        onClick={remove ? () => {} : () => onAddFilter(filter)}
        remove={remove}
      >
        {children}
      </Text>
      {remove ? (
        <Remove onClick={() => onRemoveFilter(filter)}>
          <img src="/assets/images/icon-remove.svg" alt="remove icon" />
        </Remove>
      ) : (
        ''
      )}
    </StyledSearchFilter>
  );
}

const StyledSearchFilter = styled.li`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 32px;
  margin: 10px 16px 10px 0px;
  border-radius: 5px;
  overflow: hidden; /* hides the squared border on hover of the span child */
  background-color: ${({ theme }) =>
    theme.color.neutral.lightGrayishCyanFilterTablets};

  @media (min-width: 415px) {
    margin: 10px 0px 10px 16px;
  }
`;

const Text = styled.span`
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0px 9px 0px 8px;
  color: ${({ theme }) => theme.color.primary.desaturtedDarkCyan};
  font-size: ${({ theme }) => theme.fontSize.small};
  font-weight: ${({ theme }) => theme.fontWeight.bold};

  &:hover {
    /* only apply colors if this is a not removable filter */
    color: ${({ remove, theme }) =>
      remove ? theme.color.primary.desaturtedDarkCyan : 'white'};
    background-color: ${({ remove, theme }) =>
      remove
        ? theme.color.neutral.lightGrayishCyanFilterTablets
        : theme.color.primary.desaturtedDarkCyan};
    cursor: ${({ remove }) => (remove ? 'auto' : 'pointer')};
  }
`;

const Remove = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 32px;
  height: 32px;
  border-radius: 0px 5px 5px 0px;
  background-color: ${({ theme }) => theme.color.primary.desaturtedDarkCyan};

  &:hover {
    background-color: ${({ theme }) => theme.color.neutral.veryDarkGrayishCyan};
    cursor: pointer;
  }
`;

export default SearchFilter;
