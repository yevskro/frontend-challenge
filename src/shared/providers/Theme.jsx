import React from 'react';
import { ThemeProvider } from 'styled-components';
/* theme developed based on the style-guide */
const theme = {
  color: {
    primary: {
      desaturtedDarkCyan: 'hsl(180, 29%, 50%)',
    },
    neutral: {
      lightGrayishCyanBackground: 'hsl(180, 52%, 96%)',
      lightGrayishCyanFilterTablets: 'hsl(180, 31%, 95%)',
      darkGrayishCyan: 'hsl(180, 8%, 52%)',
      veryDarkGrayishCyan: 'hsl(180, 14%, 20%)',
    },
    shadow: 'rgba(94, 140, 160, 0.15)',
  },
  fontSize: {
    xxsmall: '0.7rem',
    xsmall: '0.8rem',
    small: '0.9rem',
    medium: '1rem',
    large: '1.2rem',
  },
  fontWeight: {
    thin: '400',
    normal: '500',
    bold: '600',
  },
  fontFamily: '"Spartan", "sans-serif"',
};

const Theme = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

export default Theme;
