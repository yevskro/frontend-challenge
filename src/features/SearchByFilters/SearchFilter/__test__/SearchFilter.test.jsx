import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Theme from '../../../../shared/providers/Theme';
import SearchFilter from '..';

test('renders', () => {
  render(
    <Theme>
      <SearchFilter>CSS</SearchFilter>
    </Theme>
  );
  expect(screen.getByText('CSS')).toBeInTheDocument();
});

test('onAddFilter callback invoked', () => {
  const mock = jest.fn((filter) =>
    expect(filter).toStrictEqual({ test: 'test' })
  );

  render(
    <Theme>
      <SearchFilter onAddFilter={mock} filter={{ test: 'test' }}>
        CSS
      </SearchFilter>
    </Theme>
  );

  userEvent.click(screen.getByText('CSS'));
  expect(mock).toHaveBeenCalledTimes(1);
});

test('renders a remove filter with a remove icon', () => {
  render(
    <Theme>
      <SearchFilter value="CSS" remove />
    </Theme>
  );

  const img = screen.getByRole('img');
  expect(img).toHaveAttribute('src', '/assets/images/icon-remove.svg');
  expect(img).toBeInTheDocument();
});

test('onRemoveFilter callback invoked', () => {
  const mock = jest.fn((filter) =>
    expect(filter).toStrictEqual({ test: 'test' })
  );

  render(
    <Theme>
      <SearchFilter
        value="CSS"
        remove
        onRemoveFilter={mock}
        filter={{ test: 'test' }}
      />
    </Theme>
  );

  const img = screen.getByRole('img');
  userEvent.click(img);
  expect(mock).toHaveBeenCalledTimes(1);
});

test('snapshot', async () => {
  const mock = jest.fn();
  const tree = renderer
    .create(
      <Theme>
        <SearchFilter filter={{ test: 'test' }} onAddFilter={mock}>
          CSS
        </SearchFilter>
        <SearchFilter filter={{ test: 'test' }} remove onRemoveFilter={mock}>
          HTML
        </SearchFilter>
      </Theme>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
