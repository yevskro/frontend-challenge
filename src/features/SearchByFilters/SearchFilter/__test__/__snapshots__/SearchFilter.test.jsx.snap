// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Array [
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 32px;
  margin: 10px 16px 10px 0px;
  border-radius: 5px;
  overflow: hidden;
  background-color: hsl(180,31%,95%);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  padding: 0px 9px 0px 8px;
  color: hsl(180,29%,50%);
  font-size: 0.9rem;
  font-weight: 600;
}

.c1:hover {
  color: white;
  background-color: hsl(180,29%,50%);
  cursor: pointer;
}

@media (min-width:415px) {
  .c0 {
    margin: 10px 0px 10px 16px;
  }
}

<li
    className="c0"
  >
    <span
      className="c1"
      onClick={[Function]}
    >
      CSS
    </span>
    
  </li>,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 32px;
  margin: 10px 16px 10px 0px;
  border-radius: 5px;
  overflow: hidden;
  background-color: hsl(180,31%,95%);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  padding: 0px 9px 0px 8px;
  color: hsl(180,29%,50%);
  font-size: 0.9rem;
  font-weight: 600;
}

.c1:hover {
  color: hsl(180,29%,50%);
  background-color: hsl(180,31%,95%);
  cursor: auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 32px;
  height: 32px;
  border-radius: 0px 5px 5px 0px;
  background-color: hsl(180,29%,50%);
}

.c2:hover {
  background-color: hsl(180,14%,20%);
  cursor: pointer;
}

@media (min-width:415px) {
  .c0 {
    margin: 10px 0px 10px 16px;
  }
}

<li
    className="c0"
  >
    <span
      className="c1"
      onClick={[Function]}
    >
      HTML
    </span>
    <div
      className="c2"
      onClick={[Function]}
    >
      <img
        alt="remove icon"
        src="/assets/images/icon-remove.svg"
      />
    </div>
  </li>,
]
`;
